Light Freshness Edit: ASP.NET - Security topics (#37129)
dotnet/AspNetCore.Docs: aspnetcore/security/enforcing-ssl.md

---
title: Enforce HTTPS in ASP.NET Core
author: tdykstra
description: Learn how to require HTTPS/TLS in an ASP.NET Core web app, and find troubleshooting steps for untrusted certificate issues.
ms.author: tdykstra
monikerRange: '>= aspnetcore-3.0'
ms.custom: mvc, linux-related-content
ms.date: 05/13/2026
uid: security/enforcing-ssl

> # customer intent: As an ASP.NET Core developer, I want to enforce incoming requests to my apps to use HTTPS/TLS, so I ensure secure interaction with my app.

Reviewer: ```suggestion
# customer intent: As an ASP.NET Core web app developer, I want to force incoming requests to use HTTPS/TLS, so I can avoid insecure interaction with my apps.
```

For production environments that are implementing HTTPS for the first time, set the initial <xref:Microsoft.AspNetCore.HttpsPolicy.HstsOptions.MaxAge%2A?displayProperty=nameWithType> property value to a small amount by using one of the <xref:System.TimeSpan> methods. Set the value from hours to no more than a single day, in case you need to revert the HTTPS infrastructure to HTTP. After you're confident in the sustainability of the HTTPS configuration, increase the HSTS `max-age` value (commonly, one year).

The following highlighted code:

[!code-csharp[](enforcing-ssl/sample-snapshot/6.x/Program2.cs?highlight=7-14)]

* Sets the preload parameter of the `Strict-Transport-Security` header. Preload isn't part of the [RFC 6797 HSTS specification](https://datatracker.ietf.org/doc/html/rfc6797). Web browsers support preload of HSTS sites on fresh install. For more information, see [https://hstspreload.org/](https://hstspreload.org/).
* Enables the `includeSubDomain` directive, which applies the HSTS policy to host subdomains. For more information, see [RFC 6797 HSTS specification (Section 6.1.2)](https://datatracker.ietf.org/doc/html/rfc6797#section-6.1.2).

> * Explicitly sets the `max-age` parameter of the `Strict-Transport-Security` header to 60 days. If not set, it defaults to 30 days. For more information, see the `max-age directive` in [RFC 6797 HSTS specification (Section 6.1.1)](https://datatracker.ietf.org/doc/html/rfc6797#section-6.1.1).

Reviewer: ```suggestion
* Explicitly sets the `max-age` parameter of the `Strict-Transport-Security` header to 60 days. If not set, it defaults to 30 days. For more information, see the `max-age` directive in [RFC 6797 HSTS specification (Section 6.1.1)](https://datatracker.ietf.org/doc/html/rfc6797#section-6.1.1).
```

## Set up developer certificate for Docker

To configure the developer certificate for Docker, see [GitHub dotnet/aspnetcore.docs issue #6199](https://github.com/dotnet/AspNetCore.Docs/issues/6199) - _How to set up the dev certificate when using Docker in development_.

## Linux-specific considerations

Linux distributions differ substantially in how they mark certificates as trusted.

The `dotnet dev-certs` tool is expected to be broadly applicable, but official support is available for Ubuntu and Fedora only. The support specifically aims to ensure trust in Firefox and Chromium-based browsers (Microsoft Edge, Chrome, and Chromium).

### Dependencies

* To establish OpenSSL trust, the `openssl` tool must be on the path.
* To establish browser trust (for example in Microsoft Edge or Firefox), the `certutil` tool must be on the path.

### OpenSSL trust

> When a ASP.NET Core development certificate is trusted, the certificate is exported to a folder in the current user's home directory. To have [OpenSSL](https://www.openssl.org/) (and clients that consume it) pick up this folder, you need to set the `SSL_CERT_DIR` environment variable. You can set the variable in a single session by running a command like `export SSL_CERT_DIR=$HOME/.aspnet/dev-certs/trust:/usr/lib/ssl/certs` (the exact value is in the output when `--verbose` is passed) or by adding it your (distro- and shell-specific) configuration file (for example _.profile_).

Reviewer 2: ```suggestion
When an ASP.NET Core development certificate is trusted, the certificate is exported to a folder in the current user's home directory. To have [OpenSSL](https://www.openssl.org/) (and clients that consume it) pick up this folder, you need to set the `SSL_CERT_DIR` environment variable. You can set the variable in a single session by running a command like `export SSL_CERT_DIR=$HOME/.aspnet/dev-certs/trust:/usr/lib/ssl/certs` (the exact value is in the output when `--verbose` is passed) or by adding it your (distro- and shell-specific) configuration file (for example _.profile_).
```

>[!NOTE]
> Remove the curl alias you created earlier

-->

## Troubleshoot certificate problems (certificate not trusted)

> Sometimes when a ASP.NET Core HTTPS development certificate is [installed and trusted](#trust), the browser warns that the certificate is untrusted. The following sections provide help for troubleshooting this issue.

Reviewer 2: ```suggestion
Sometimes when an ASP.NET Core HTTPS development certificate is [installed and trusted](#trust), the browser warns that the certificate is untrusted. The following sections provide help for troubleshooting this issue.
```
